### PR TITLE
ext_time_quota_acl: remove -l option

### PR DESCRIPTION
**Pull request by kinkie**
Supporting logging to a file complicates upgrading helper code to use
debugs() because DebugFile code calls commSetCloseOnExec(), and our
comm/libminimal does not currently provide a functioning implementation
for that API: The existing implementation is an unconditional assert.
To save development time while upgrading helpers, we are dropping this
feature. It can probably be emulated using shell redirection tricks.

### CODE REVIEW

**Comment by kinkie** [comment]
This is a stepping stone for pr #1847 which will hopefully immediately follow

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (rousskov) [inline src/acl/external/time_quota/ext_time_quota_acl.cc:81]: > PR description: Debug logging to own file does not work well when there are
multiple helpers, which could clobber each other's output.

I have re-checked what fopen("a") means. The above assertion is rather misleading because fopen("a") guarantees no clobbering by multiple processes writing debugging messages as long as each underlying write(2) call is given the whole message (i.e. no partial write(2)s due to fwrite() buffering). If all actual debugging messages in this helper are smaller than fwrite() buffers, then there will be no clobbering. I speculate that they are smaller in supported environments.

> It's better to log through squid and find the output in cache.log.

Squid's cache.log is shared among multiple SMP kids. Its code uses exactly the same fopen("a") approach. Thus, if helpers do have clobbering problems in some environments, then migrating to cache.log will make those clobbering problems _worse_ in SMP Squid cases!

I suggest dropping the clobbering claim and appealing to debugs() upgrade overheads instead. I edited PR description accordingly. Please review and undo or adjust as needed!

P.S. I have not yet fully checked what it would take to solve the commSetCloseOnExec() problem. I am currently relying exclusively on your assessment that there is no trivial solution there.

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (rousskov) [inline src/acl/external/time_quota/ext_time_quota_acl.cc:424]: This PR should update release notes, but, to save time, you may want to wait for @yadij opinion regarding this feature removal before doing so.

Comment (kinkie) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:424]: I thought about it; given the fact that it is not a major feature
I considered just recording it to the ChangeLog when this lands.
I'll also defer to Amos' judgement if he feels this deserves go to
the release notes

Comment (rousskov) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:424]: Sounds good. I am OK with whatever you two can agree on with regard to release notes inclusion.

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (kinkie) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:81]: > fopen("a") guarantees no clobbering by multiple processes writing debugging messages as long as each underlying write(2)

The MacOS manual page reads:
```
     “a”     Open for writing.  The stream is positioned at the end of the file.
             Subsequent writes to the file will always end up at the then
             current end of file, irrespective of any intervening fseek(3) or
             similar.  Create the file if it does not exist.
```
In other words, it guarantees that each write happens at the end of file, but I can't find any mention of the guarantee you mention.
In a worst-case scenario:

1. we have no guarantee that printf() will issue one call to write() per call (although it's likely, but probably implementation-dependent)
2. we have no guarantee that a call to write() will not be interrupted

Let's imagine a scenario where we have two processes writing to the same file in "a" mode.
* A issues its command, writes some output but to complete the write the OS requires allocating a block on storage. A gets suspended while the OS does that
* After the OS allocates the storage but before A gets a chance to run again, B gets a chance to run. B has a pending write, and issues it, just in the middle of A's output
* A now gets a chance to finish its write operation. It does so at the end of file, without overwriting what B has just written to disk, but now the last two entries written to the file are corrupted.

Probably not a big deal, but a possible scenario.

Squid suffers from the same problem, although probably to a lesser extent: I expect that a typical installation will have many more helpers than squid worker processes running, as most helpers do not support multithreaded operations and more processes are needed to match squid's parallelism

Comment (rousskov) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:81]: Are you _sure_ we should be spending time on this? If you are, I will find the time to clarify and probably invalidate the proposed scenario. As I mentioned earlier, I recall being satisfied when I studied the issue at depth several years ago. My current recollection may be faulty, of course.

> Squid suffers from the same problem, although probably to a lesser extent

Modern busy SMP Squids in ALL,9 debugging mode would probably be much, much more susceptible to this problem if it were real. FWIW, I have never seen them clobber their (often large) debugs() messages. I could miss some rare cases, of course. However, my prior research was not based on these observations. It was based on standards and their implementations.

Comment (rousskov) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:81]: If you want a bit more information for making that "spend more time on this" decision, consider reading [Linus](https://lkml.iu.edu/hypermail/linux/kernel/0208.0/0251.html) and [others](https://stackoverflow.com/questions/12942915/understanding-concurrent-file-writes-from-multiple-processes#comment94938866_12943431) (while carefully filtering out noise and irrelevant details).

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (yadij) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:424]: This is a UI feature, so IMO it should have some release notes mention. Does not have to be a large mention, just something for us to point at when people get surprised that their custom logs no longer work.

**Review by yadij — APPROVED** [review]

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (kinkie) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:424]: added. Please re-review

File: doc/release-notes/release-7.sgml.in
Comment (rousskov) [inline doc/release-notes/release-7.sgml.in:127]: For consistency with other `<sect1>` titles above:

```suggestion
<sect1>Removed custom log location for ext_time_quota_acl helper
```

I would also clarify with:

```suggestion
<sect1>Removed custom log location support from ext_time_quota_acl helper
```

... and even generalize since we mention formatting changes:

```suggestion
<sect1>Adjusted configuration and format of ext_time_quota_acl helper debugging
```

Comment (kinkie) [reply doc/release-notes/release-7.sgml.in:127]: adopting the third option

**Review by yadij — DISMISSED** [review]

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (kinkie) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:424]: Adding a default clause will do . Doing that

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (kinkie) [inline src/acl/external/time_quota/ext_time_quota_acl.cc:400]: Removed this line; it is misleading.
usage() gets invoked when the -h option is used; it's not a wrong usage.

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (kinkie) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:81]: Got it.
If one were to pick nits, `DebugChannel::writeToStream` uses fprintf(3),
not write(2). Given the short size of the output, it probably doesn't make a difference. I wonder if there would be benefit by using writev() instead.

Comment (rousskov) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:81]: > If one were to pick nits,

... then one should not ignore the word "underlying" in "underlying write(2)" in the [original statement](https://github.com/squid-cache/squid/pull/1872#discussion_r1690275970).

> Given the short size of the output, it probably doesn't make a difference.

Yes, in typical environments (at least), stdio buffers are evidently large enough for typical cache.log messages. I may have even checked what the defaults are and found them satisfactory when I was researching this, but I am not sure.

> I wonder if there would be benefit by using writev(2) instead.

Most likely, we want to get rid of the remaining printf() and such in debugging stream implementation. Once that is done, we will probably just write(2) a single buffer (filled by in-memory std::ostream writes a la ToSBuf()) -- no need to build a vector for writev(2).

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (rousskov) [inline src/acl/external/time_quota/ext_time_quota_acl.cc:423]: This code does not follow getopt(3) API AFAICT -- printed `opt` does _not_ store the name of an unknown option in this context. I have not checked all of the details, but I would start with testing getopt(3) build-in error reporting:

```suggestion
```

or

```suggestion
            // getopt() emits error message to stderr
```

If that does not work, look for "getopt() can detect two kinds" in getopt(3) [documentation](https://www.man7.org/linux/man-pages/man3/getopt.3.html).

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/acl/external/time_quota/ext_time_quota_acl.cc
Comment (kinkie) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:423]: Done

Comment (rousskov) [reply src/acl/external/time_quota/ext_time_quota_acl.cc:423]: > Done.

Not done (including/or not pushed).

Comment (kinkie) [reply src/acl/external/time_quota/ext_time_quota_acl.cc]: Sigh.. sorry.
Now done for real

**Comment by kinkie** [comment]
All outstanding requests for change are addressed. Waiting for approval or slow burner timeout, clearing for merge in the meanwhile

**Comment by kinkie** [comment]
Hand-backporting in PR #1909

**Pull request closed without merging**
